silverback/runner.py

```diff
@@ -181,6 +188,12 @@ async def run(self):
                 await asyncio.gather(*(self.recorder.add_result(r) for r in converted_results))
 
             # NOTE: No need to handle results otherwise
+        # NOTE: Do one last backup
+        backup_task = await self.app.backup_task.kiq(self.state)
+        result = await backup_task.wait_result()
```

Comment:
question: Isn't `self.state` worker-only?

Reply:
I guess this is runner state, but a worker will execute the save via the data store?  Why give a worker this responsibility?

Reply:
> question: Isn't `self.state` worker-only?

Yes it is, I have uncommited work that completely removes `Runner.state` (and also tracking `last_block_..` as a Runner instance value)

Reply:
> I guess this is runner state, but a worker will execute the save via the data store? Why give a worker this responsibility?

Was using this just to track `last_block_....` but then realized why does the runner have any state at all? track it all via worker (and backup via worker too)

Reply:
There's a few reasons I think it makes more sense in the runner.  

- The runner is the one handling chain data and initiating jobs.  It knows what blocks were last seen and plopped onto the queue.  When the runner comes back up, it will initiate the startup job notifying the workers allowing them to do any back-processing (the original use case of this block tracking feature)
- In a multi-worker (or multi-app) environment, there's one runner.  So we either have one runner acting as source of truth, or every worker with their own source of truth potentially having some synchronization issues
- The centralized runner in a distributed system already has state handling and data access
- Workers run user code, so we'll essentially have to grant user code access to this data

Reply:
So far, the runner is still initiating the backup by supplying the blocks, I think given what you said that makes sense

However, application state has to live in the application. The backup would have to round-trip the snapshot back to the runner in order to perform the backup fully on the runner side every time it wishes to make a backup (before and after every single continuous task the runner triggers)

That could be done, but I think a bit of tradeoff there is that there would be more heavy broker communication if application state was very big to do a snapshot, whereas if the worker is the one initiating the snapshots, that overhead is a static size. The backup system (CCS?) sees the same loads either way.

Reply:
I wouldn't want to add any bi-directional communication between runner and bot.  With the exception of results, I think everything should be on-way.  I'm mostly concerned about separation of concerns and synchronization.

Guess I'll just see what happens when I try and integrate this.

Reply:
Yeah I think this only really works if the only time the state is loaded is on app startup (after worker startup, where `app.state` is still not set yet), and the rest of the time it only exports snapshots from the worker on demand (as triggered by the runner as a system task)